src/Visualization/Python/PlotSizeControl.py

```diff
@@ -0,0 +1,361 @@
+    import numpy as np
+
+    if ax is None:
+        ax = plt.gca()
+    newlow, newhigh = np.inf, -np.inf
+
+    for artist in ax.collections + ax.lines:
+        x, y = _get_xy(artist)
+        if axis == "y":
+            setlim = ax.set_ylim
+            lim = ax.get_xlim()
+            fixed, dependent = x, y
+        else:
+            setlim = ax.set_xlim
+            lim = ax.get_ylim()
+            fixed, dependent = y, x
+
+        low, high = _calculate_new_limit(fixed, dependent, lim)
+        newlow = low if low < newlow else newlow
+        newhigh = high if high > newhigh else newhigh
+
+    margin = margin * (newhigh - newlow)
+
+    setlim(newlow - margin, newhigh + margin)
+
+
+@click.command(name="plot-size-control")
+@click.argument(
+    "reduction_files",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, readable=True),
+    nargs=-1,
+)
+@click.option(
+    "--object-label",
+    "-d",
+    required=True,
+    help=(
+        "Which object to plot. This is either 'A', 'B', or 'None'. 'None' is"
+        " used when there is only one black hole in the simulation."
+    ),
+)
+@click.option(
+    "--output",
+    "-o",
+    type=click.Path(file_okay=True, dir_okay=False, writable=True),
+    help=(
+        "Name of the output image. The '--object-label' will be added"
+        " automatically to the end of the output. Also the suffix '.pdf' will"
+        " be added if necessary. If unspecified, the plot is shown"
+        " interactively, which only works on machines with a window server."
+    ),
+)
+# Plotting options
+@click.option(
+    "--xlim",
+    type=float,
+    nargs=2,
+    help="The lower and upper bounds of the x-axis.",
+)
+@click.option(
+    "--xlabel",
+    help="The label on the x-axis.",
+)
+@click.option(
+    "--title",
+    "-t",
+    help="Title of the graph.",
+)
+@click.option(
+    "--stylesheet",
+    "-s",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, readable=True),
+    envvar="SPECTRE_MPL_STYLESHEET",
+    help=(
+        "Select a matplotlib stylesheet for customization of the plot, such "
+        "as linestyle cycles, linewidth, fontsize, legend, etc. "
+        "The stylesheet can also be set with the 'SPECTRE_MPL_STYLESHEET' "
+        "environment variable."
+    ),
+)
+def plot_size_control_command(
+    reduction_files: Sequence[str],
+    object_label: str,
+    output: str,
+    xlim: Sequence[float],
+    xlabel: str,
+    title: str,
+    stylesheet,
+):
+    """
+    Plot diagnostic information regarding the Size control system.
+
+    This tool assumes there is a subfile in each of the "reduction-files" with
+    the path `/ControlSystems/Size{LABEL}/Diagnostics.dat`, where `{LABEL}` is
+    replaced with the "object-label" input option.
+    """
+    _rich_traceback_guard = True  # Hide traceback until here
+
+    object_label = (
+        "" if object_label.lower() == "none" else object_label.upper()
+    )
+
+    times = np.array([])
```

Comment:
This is actually almost trivial with pandas:

```py
import pandas as pd
from spectre.Visualization.ReadH5 import to_dataframe

data = pd.concat(to_dataframe(h5py.File(filename)[subfile_name]) for filename in filenames)

# That's all! Replaces these ~90 lines.
```

Reply:
Note: if the amount of data becomes too large you can also do `to_dataframe(..., slice=np.s_[::step])` to load only every `step` observation.